File: backend/app/schemas/employer.py
"""Employer domain Pydantic schemas for dashboard, applicants, and job management."""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.application import ApplicationStatus
from app.models.job import JobStatus


class EmployerDashboardMetrics(BaseModel):
    """Aggregated numerical metrics for employer portal."""

    total_jobs: int = 0
    published_jobs: int = 0
    draft_jobs: int = 0
    closed_jobs: int = 0
    total_applications: int = 0
    applications_by_status: dict[str, int] = Field(default_factory=dict)


class EmployerRecentJobItem(BaseModel):
    """Snapshot of a recent job requisition for dashboard display."""

    id: uuid.UUID
    title: str
    status: JobStatus
    location_city: str | None = None
    applications_count: int = 0
    skills_count: int = 0
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EmployerRecentApplicationItem(BaseModel):
    """Snapshot of a recent candidate application for dashboard display."""

    id: uuid.UUID
    candidate_id: uuid.UUID
    candidate_name: str
    candidate_headline: str | None = None
    job_id: uuid.UUID
    job_title: str
    status: ApplicationStatus
    applied_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EmployerDashboardResponse(BaseModel):
    """Comprehensive real-time dashboard data payload."""

    metrics: EmployerDashboardMetrics
    recent_jobs: list[EmployerRecentJobItem] = Field(default_factory=list)
    recent_applications: list[EmployerRecentApplicationItem] = Field(default_factory=list)


class EmployerApplicationCandidateInfo(BaseModel):
    """Safe candidate profile and identity details presented to employers."""

    id: uuid.UUID
    full_name: str
    email: str
    headline: str | None = None
    bio: str | None = None
    experience_years: float = 0.0
    education_level: str | None = None
    location_city: str | None = None
    location_state: str | None = None

    model_config = ConfigDict(from_attributes=True)


class EmployerApplicationResponse(BaseModel):
    """Detailed job application representation for employer review."""

    id: uuid.UUID
    candidate_id: uuid.UUID
    job_id: uuid.UUID
    job_title: str
    status: ApplicationStatus
    cover_note: str | None = None
    applied_at: datetime
    candidate: EmployerApplicationCandidateInfo

    model_config = ConfigDict(from_attributes=True)


class EmployerApplicationStatusUpdate(BaseModel):
    """Payload for updating candidate job application status."""

    status: ApplicationStatus

File: frontend/src/types/employer.ts
export type EmploymentType = "FULL_TIME" | "PART_TIME" | "CONTRACT" | "INTERNSHIP";
export type ExperienceLevel = "ENTRY" | "MID" | "SENIOR" | "LEAD";
export type JobStatus = "DRAFT" | "PUBLISHED" | "CLOSED";
export type ApplicationStatus =
  | "APPLIED"
  | "SHORTLISTED"
  | "INTERVIEW"
  | "OFFERED"
  | "REJECTED"
  | "HIRED";

export interface JobSkillRequirement {
  skill_id: string;
  is_required: boolean;
  minimum_proficiency: string;
  weight: number;
}

export interface JobSkill {
  id: string;
  skill_id: string;
  skill_name?: string | null;
  category?: string | null;
  is_required: boolean;
  minimum_proficiency: string;
  weight: number;
}

export interface Job {
  id: string;
  employer_id: string;
  employer_name?: string | null;
  title: string;
  description: string;
  location_city?: string | null;
  location_state?: string | null;
  is_remote: boolean;
  employment_type: EmploymentType;
  experience_level: ExperienceLevel;
  status: JobStatus;
  salary_min?: number | null;
  salary_max?: number | null;
  is_active: boolean;
  skills: JobSkill[];
  applications_count: number;
  created_at: string;
  updated_at: string;
}

export interface JobCreatePayload {
  title: string;
  description: string;
  location_city?: string | null;
  location_state?: string | null;
  is_remote?: boolean;
  employment_type?: EmploymentType;
  experience_level?: ExperienceLevel;
  status?: JobStatus;
  salary_min?: number | null;
  salary_max?: number | null;
  skills?: JobSkillRequirement[];
}

export interface JobUpdatePayload {
  title?: string;
  description?: string;
  location_city?: string | null;
  location_state?: string | null;
  is_remote?: boolean;
  employment_type?: EmploymentType;
  experience_level?: ExperienceLevel;
  status?: JobStatus;
  salary_min?: number | null;
  salary_max?: number | null;
  is_active?: boolean;
  skills?: JobSkillRequirement[];
}

export type { Skill } from "./skill";

export interface CandidateInfo {
  id: string;
  full_name: string;
  email: string;
  headline?: string | null;
  bio?: string | null;
  experience_years: number;
  education_level?: string | null;
  location_city?: string | null;
  location_state?: string | null;
}

export interface EmployerApplication {
  id: string;
  candidate_id: string;
  job_id: string;
  job_title: string;
  status: ApplicationStatus;
  cover_note?: string | null;
  applied_at: string;
  candidate: CandidateInfo;
}

export interface EmployerDashboardMetrics {
  total_jobs: number;
  published_jobs: number;
  draft_jobs: number;
  closed_jobs: number;
  total_applications: number;
  applications_by_status: Record<string, number>;
}

export interface EmployerRecentJob {
  id: string;
  title: string;
  status: JobStatus;
  location_city?: string | null;
  applications_count: number;
  skills_count: number;
  created_at: string;
}

export interface EmployerRecentApplication {
  id: string;
  candidate_id: string;
  candidate_name: string;
  candidate_headline?: string | null;
  job_id: string;
  job_title: string;
  status: ApplicationStatus;
  applied_at: string;
}

export interface EmployerDashboardData {
  metrics: EmployerDashboardMetrics;
  recent_jobs: EmployerRecentJob[];
  recent_applications: EmployerRecentApplication[];
}

export interface EmployerProfile {
  id: string;
  user_id: string;
  company_name: string;
  company_description?: string | null;
  industry?: string | null;
  location_city?: string | null;
  location_state?: string | null;
  website_url?: string | null;
  created_at: string;
  updated_at: string;
}

export interface EmployerProfileUpdatePayload {
  company_name?: string;
  company_description?: string | null;
  industry?: string | null;
  location_city?: string | null;
  location_state?: string | null;
  website_url?: string | null;
}
